src/transformers/trainer.py

```diff
@@ -1126,13 +1137,20 @@ def get_optimizer_cls_and_kwargs(
                 The training arguments for the training session.
 
         """
+        use_badam = "badam_" in args.optim
+        if use_badam:
+            args.optim = args.optim.replace("badam_", "")
 
         # parse args.optim_args
         optim_args = {}
+        badam_optim_args = {}
         if args.optim_args:
             for mapping in args.optim_args.replace(" ", "").split(","):
                 key, value = mapping.split("=")
-                optim_args[key] = value
+                if "badam_" in key and use_badam:
```

Comment:
nit - check the bool flag first in the and check, it'll be faster 

```suggestion
                if use_badam and "badam_" in key:
```